This is to fix issue #786 - conversion of string/number/bool is

cwikla opened the pull request:
different for optional/non-optional

stringValue now just calls string, and passes back the default "" if
it received nil

numberValue now just calls number, and passes back the default
NSNumber(0.0) if it received nil

boolValue now just calls bool, and passes back the default false
if it received nil

Added support for "yes" and "true" in bool conversion since "y" and "t"
were accepted.

Updated Number and String tests to match

cwikla commented:
No interest on this pull request? No comments?  I'd fix more stuff but not if it doesn't go anywhere.

kengruven commented:
There's a few things going on here.  One change is to add support for the string "yes" as a bool.  On its own, that looks reasonable to me.

The bigger change, to modify the meaning of the getters, is less clear to me.

First, there's some easy fixes which would probably make this more likely to be considered:
 - fix whitespace (as reported by the bot)
 - use nil coalescing, rather than "if let" (easier to read, and consistent with other getters)
 - update documentation to accurately describe the new functionality

However, it's not clear to me that this is a worthwhile change to make.  The `if let val = json.typename` idiom is very convenient, and if we made all getters disregard actual types and just try their best, we'd lose this ability.  In my experience, being able to pass a number in JSON as a number-or-string just isn't a big use case, and not worth making each JSON access check the type on its own.

cwikla commented:
Yes as bool is cute, doesn't really matter.

I fixed the white space problem 10 months ago, probably all broke now.

For the nil coalescing - I literally moved existing code around and added wrappers to
make each case work the same. So that's a matter of style that may have matured in the last 10 months.

Documentation I feel is currently broken since this should be the expected behavior.

I think your experience is wrong.  First time using this code it did unexpected things (see original example) - which I think if you ask 10 of your fellow engineers if they don't think it should work that way, I'd be surprised.  If this is supposed to be the case I will claim the documentation should be updated to show that it is intended behavior because it is unexpected.
